### PR TITLE
[frieren] Round-2: FiLM + projection + protocol full composition stack

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

All three independent Round-1 wins on `yi` were measured in isolation. This experiment stacks them in a single run to verify they compose additively — and to push `abupt_axis_mean` below 14.

The three wins and their measured standalone deltas:
1. **Per-block FiLM geometry conditioning** (your PR #8): 30.47 → 16.53 on same 1-epoch config (−46% vs no-FiLM)
2. **Volume-loss reweight + protocol fixes** (PR #9): 35.12 → 17.39 (−50%) — driven mainly by bs=8 + validation-every=1
3. **Tangential wall-shear projection** (PR #11): first yi baseline, confirms physics loss is additive

None of these three wins overlaps the other mechanistically:
- FiLM adds a per-geometry conditioning pathway
- Protocol fixes give more gradient updates per epoch
- Projection removes physically unphysical loss signal

If they compose multiplicatively (unlikely but possible), we'd reach ~7. Even conservative additive composition should clear 13–14. This is the most important experiment in Round 1.

## Instructions

**Step 1: Rebase your FiLM branch onto `yi`**

Your PR #8 is currently blocked on a rebase conflict. Finish the rebase first:

```bash
git fetch origin
git checkout frieren/round1-geometry-film-conditioning
git rebase origin/yi
# resolve any conflicts in train.py, then:
git push --force-with-lease
```

Once PR #8 is merged, the FiLM code will be on `yi`. Then:

```bash
git checkout frieren/film-protocol-composition-stack
git rebase origin/yi
```

**Step 2: Your implementation is already complete.** The FiLM conditioning is in your PR #8 code (now on `yi`). No new code is needed. This PR only changes the CLI flags.

**Step 3: Run the composition experiment:**

```bash
python train.py \
  --volume-loss-weight 2.0 \
  --batch-size 8 \
  --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --use-tangential-wallshear-loss \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms \
  --wandb-group film-composition-stack
```

The only additions over the PR #9 winning config are:
- `--use-tangential-wallshear-loss` (PR #11 code on `yi`)
- FiLM conditioning is active if your PR #8 code is in `train.py`

**Step 4: If PR #22 (gradient clipping) has merged before you run, add it.**
PR #22 (gilbert) adds `torch.nn.utils.clip_grad_norm_(model.parameters(), max_norm=1.0)` between `loss.backward()` and `optimizer.step()`. This prevents the loss spikes seen in several Round-1 runs. If it's on `yi` when you rebase, your run will inherit it automatically.

## Baseline

Current yi best (PR #9 gilbert, run `y2gigs61`, 6 epochs, best_epoch=3):

| Metric | yi best (PR #9) | FiLM alone (PR #8 pending) | AB-UPT target |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | 17.39 | **16.53** | — |
| `surface_pressure_rel_l2_pct` | 11.07 | 10.38 | 3.82 |
| `wall_shear_rel_l2_pct` | 18.32 | 17.29 | 7.29 |
| `volume_pressure_rel_l2_pct` | 15.21 | 14.91 | 6.08 |
| `wall_shear_x_rel_l2_pct` | 15.65 | 14.76 | 5.35 |
| `wall_shear_y_rel_l2_pct` | 21.86 | 20.59 | 3.65 |
| `wall_shear_z_rel_l2_pct` | 23.18 | 22.00 | 3.63 |

Beat 16.53 on `abupt_axis_mean_rel_l2_pct` to set a new yi best.

**Reproduce (PR #9 base config):**
```bash
cd target/
python train.py \
  --volume-loss-weight 2.0 --batch-size 8 --validation-every 1 \
  --lr 2e-4 --weight-decay 5e-4 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 --model-slices 128 \
  --ema-decay 0.9995 \
  --gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms
```

## What to report

Post your W&B run ID(s) and the full `test_primary/*` table in a comment. If gradient clipping was present, note it. Report any train→val divergence and what epoch best_val was recorded.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: **Yi new best — PR #13 progressive EMA anneal verified (pending merge).**

W&B run `wio9pqw2` confirmed: `test_primary/abupt_axis_mean = 15.82` — a **9% improvement** over the current official yi baseline of 17.39 (PR #9). Merge is blocked on a rebase, but results are accepted. Treat 15.82 as your new comparison target.

**Updated reference table:**

| Metric | New yi best (PR #13, pending) | Official yi (PR #9 merged) | AB-UPT target |
|---|---:|---:|---:|
| `abupt_axis_mean_rel_l2_pct` | **15.82** | 17.39 | — |
| `surface_pressure_rel_l2_pct` | **9.99** | 11.07 | 3.82 |
| `wall_shear_rel_l2_pct` | **16.60** | 18.32 | 7.29 |
| `volume_pressure_rel_l2_pct` | **14.21** | 15.21 | 6.08 |
| `wall_shear_x_rel_l2_pct` | **14.27** | 15.65 | 5.35 |
| `wall_shear_y_rel_l2_pct` | **19.49** | 21.86 | 3.65 |
| `wall_shear_z_rel_l2_pct` | **21.12** | 23.18 | 3.63 |

**What norman did:** Cosine-annealed `--ema-decay` from 0.99 (start) to 0.9999 (end) using two new flags: `--ema-decay-start 0.99 --ema-decay-end 0.9999`. Used the PR #9 winning base config (vol_w=2.0, bs=8, validation-every=1). No new parameters. 4 epochs, val improved monotonically — no train→val divergence.

**Action for your PR:**

1. **Add `--ema-decay-start 0.99 --ema-decay-end 0.9999` to your run command** — this code will be on `yi` once PR #13 merges (likely today). If you need it now, you can cherry-pick from `norman/round1-progressive-ema-decay`.

2. **Critical revision of the (B) divergence verdict:** PR #20 (nezuko EMA sweep) diagnosed post-epoch-1 divergence in a small-batch config. PR #13 shows that with `--batch-size 8 --volume-loss-weight 2.0` the divergence **does not occur** — val improved through epoch 4. If your PR uses bs=8 + vol_w=2.0, do not preemptively cut epochs assuming divergence.

3. **Rebase onto `yi`** if you have not already, to pick up all cumulative fixes (`af92e9a` per-step timeout + projection code + cosine EMA once merged).

4. **Continue using:** `--gradient-log-every 100 --weight-log-every 100 --no-log-gradient-histograms`

5. **Gradient clipping bug remains:** `train.py` still has no `clip_grad_norm_`. PR #22 (gilbert) will add it. If you see loss spikes, note it — but with bs=8 + vol_w=2.0 it may not bite.

**Note for frieren PR #8 (FiLM) and PR #23 (composition):** FiLM (16.53 standalone) is now superseded as a standalone best by cosine EMA (15.82), but both are orthogonal — one is checkpointing, one is architecture conditioning. Stacking all four wins (FiLM + projection + protocol + cosine EMA) should push well below 14.

— ADVISOR

**Comment by morganmcg1** [comment]
OPERATOR: Pausing this follow-up assignment temporarily so student:frieren can first resolve the active dependency PR #8 rebase/merge blockage. Removing student:frieren from this PR; the advisor should reassign this composition follow-up after #8 lands on yi.

**Comment by morganmcg1** [comment]
OPERATOR UPDATE: Using the cleaner paused state for this follow-up: keeping student:frieren for ownership, but removing status:wip so it is not picked up until PR #8 lands.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #23 because Closing this draft — it was designed against the PR #9 baseline (17.39). The current yi best is 12.74 (PR #66). The composition hypothesis (FiLM + projection + protocol) has since been superseded by the full 6L/256d stack with per-axis loss upweighting. The FiLM code landed via PR #8 and the projection via PR #11; both are on yi. If frieren wants to revisit a composition experiment, it should be designed against the Round-3 baseline with the full current config..
